### PR TITLE
PERF: RangeIndex.__getitem__ with integers return RangeIndex

### PR DESCRIPTION
**Pull request by mroeschke**
Builds on #57752

Discovered in #57441

### CODE REVIEW

File: pandas/core/indexes/range.py [hunk 1]
```diff
@@ -897,12 +910,19 @@ def symmetric_difference(
             result = result.rename(result_name)
         return result
 
+    def _join_empty(
+        self, other: Index, how: JoinHow, sort: bool
+    ) -> tuple[Index, npt.NDArray[np.intp] | None, npt.NDArray[np.intp] | None]:
+        if other.dtype.kind == "i":
```
Comment (WillAyd) [inline pandas/core/indexes/range.py:916]: Why is this specific to just this dtype? I am somewhat wary of doing this one-off in a function like this

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
This looks to have a lot of overlap with https://github.com/pandas-dev/pandas/pull/57752 - is that expected?

**Comment by mroeschke** [comment]
> This looks to have a lot of overlap with #57752 - is that expected?

Yes, this relies on some of the changes in https://github.com/pandas-dev/pandas/pull/57752

File: pandas/core/indexes/range.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/range.py:916]: We can only convert `Index[int]` to `RangeIndex` (which what `_shallow_copy` will do if possible) to use later in `_join_empty` to ensure that returns a `RangeIndex`

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -472,18 +472,31 @@ def _shallow_copy(self, values, name: Hashable = no_default):
...
             return Index(values, name=name, dtype=np.float64)
-        if values.dtype.kind == "i" and values.ndim == 1 and len(values) > 1:
+        if values.dtype.kind == "i" and values.ndim == 1:
             # GH 46675 & 43885: If values is equally spaced, return a
             # more memory-compact RangeIndex instead of Index with 64-bit dtype
+            if len(values) == 0:
+                return type(self)._simple_new(_empty_range, name=name)
+            elif len(values) == 1:
+                start = values[0]
+                new_range = range(start, start + self.step, self.step)
```
Comment (WillAyd) [inline pandas/core/indexes/range.py:482]:
```suggestion
                new_range = range(start, start + self.stop, self.step)
```

I think this is a typo?

Comment (mroeschke) [reply pandas/core/indexes/range.py:482]: Nope this was intentional. We want to create a new length 1 `RangeIndex` with the start of `values` and the step of `self`

**Review by WillAyd — APPROVED** [review]

**Pull request merged**